sapicore/engine/synapse/STDP.py

```diff
@@ -0,0 +1,161 @@
+
+        self.mu_plus = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + mu_plus
+        self.mu_minus = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + mu_minus
+
+        self.alpha_plus = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + alpha_plus
+        self.alpha_minus = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + alpha_minus
+
+        # state attributes specific to STDP synapses, over and above Synapse.
+        self.src_last_spiked = torch.zeros(self.matrix_shape[1], dtype=torch.float, device=self.device)
+        self.dst_last_spiked = torch.zeros(self.matrix_shape[0], dtype=torch.float, device=self.device)
+
+        # toggle learning switch (should be on during training and off during testing).
+        self.learning_switch = True
+
+    def update_weights(self) -> Tensor:
+        """STDP weight update implementation.
+
+        Returns
+        -------
+        Tensor
+            2D tensor containing weight differences (dW).
+
+        """
+        # update last spike time to current simulation step iff a spike has just arrived at the synapse.
+        self.src_last_spiked = self.src_last_spiked + self.delayed_data * (-self.src_last_spiked + self.simulation_step)
+
+        # update last spike time only for postsynaptic units that spiked in this model iteration.
+        self.dst_last_spiked = self.dst_last_spiked + self.dst_ensemble.spiked * (
+            -self.dst_last_spiked + self.simulation_step
+        )
+
+        # initialize dW matrix--remember the format is dst.num_units by src.num_units.
+        delta_weight = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device)
+
+        # subtract postsynaptic last spike time stamps from presynaptic.
+        # transpose dst_last_spiked to a column vector, extend column-wise, then add its negative to src_last_spiked.
+        dst_tr = self.dst_last_spiked.reshape(self.dst_last_spiked.shape[0], 1)
+        delta_spike = -dst_tr.repeat(dst_tr.shape[1], self.src_ensemble.num_units) + self.src_last_spiked
+
+        # spike time differences for the potentiation and depression cases (pre < post, pre > post, respectively).
+        ltp_diffs = (delta_spike < 0.0).int() * delta_spike
+        ltd_diffs = (delta_spike > 0.0).int() * delta_spike
+
+        # add to total delta weight matrix (diffs are in simulation steps, tau are in ms).
+        delta_weight = delta_weight + (delta_spike < 0.0).int() * (
+            self.alpha_plus * torch.exp(ltp_diffs / (self.tau_plus / self.dt))
+        )
+        delta_weight = delta_weight + (delta_spike > 0.0).int() * (
+            -self.alpha_minus * torch.exp(-ltd_diffs / (self.tau_minus / self.dt))
+        )
+
+        self.weights = self.weights.add(delta_weight)
+        return delta_weight
+
+    def toggle_learning(self, state: bool = False) -> None:
```

Comment:
Toggle gives the impression that it is a toggle button (flips it from the last state). It should be named something like `set_learning`

Reply:
Also, the variable should just be called `learning` or `is_learning`.